### PR TITLE
#158, #143 allow custom date validation, tweak styles

### LINKED ISSUE #143: Adjust width of date part fields
Can the width of the date part fields be configurable? Alternatively, can the month field (text) be wider and the day and year fields (numeric, 4 digits max) be narrower? Some fonts are not able to display the full month name in the current fixed width. Thanks!

### LINKED ISSUE #158: Allow custom min/max/etc. and messages for date CustomDataTypes
Our ThreePartsDate and BirthDate may now be able to have some custom validation using the `parameters` or `code_parameters` attributes of the [CustomDataType class](https://docassemble.org/docs/fields.html#custom%20datatype). For example, some kind of `min` and `max` values (`almin`, `almax`?) which could be python code.

We'd have to specify to users that it would have to be python code because we have to choose between a string value or code value when we make the parameters. Also, I can't yet find a way to automatically parse Mako.

Another aspect is allowing custom validation messages. I think those would need to be their own "top level" keys/directives because of where we can access those values in the custom data type class. For example:

```yml
fields:
  - "Birtdate": users[0].birthdate
    datatype: BirthDate
    almin: 01/01/0001
    alminmessage: "The answer needs to be an AD date"
```

Whether we can somehow work in custom validation code is a different question. From what I can tell so far, evaluating that client-side would involve using `eval()`, which is unsafe to call on arbitrary code. I'm not sure how we'd get the value server-side, though, given the tools we have.

### PR DESCRIPTION
**Pull request by plocket**
Closes #158 , closes #143 

- Allow developers to add min and max dates.
- Allow developers to add custom validation messages.
- Styles: Give more room for the text of the month field to show.
- Attempt to add better accessibility features for the inputs. Tested with a mac screen reader, but I found it somewhat hard to interpret what the screen reader was saying.

The first 3 commits are pretty organized into chunks if you want to see changes separately. They are somewhat large chunks. The commit messages should clue you into ones that have non-cosmetic changes.

When there are multiple invalid inputs to a field, the visible message doesn't always align with the field in which the cursor is focused.

There are some bugs on submit that I noted. We can turn those into issues. I think they may require upstream work, though.

I outlined behavioral requirements, and also included notes that are in the code about bugs, todos, limitations, etc, in [this comment](https://github.com/SuffolkLITLab/docassemble-ALToolbox/issues/158#issuecomment-1465239177).

### CODE REVIEW

**Review by BryceStevenWilley — CHANGES_REQUESTED** [review]
We tested it a bit more, and did come across a bug that is blocking; if a date input isn't filled in, javascript will break on the page, and none of the widgets will load. It's pre-existing, and we can trigger the same issue on the main branch, but it's a bit easier to trigger the issue in this PR.

We should at least know a bit more about what causes the above bug and if it needs an upstream fix or if we can fix it in our code before merging.

File: setup.py [hunk 1]
```diff
@@ -46,15 +46,15 @@ def find_package_data(where='.', package='', exclude=standard_exclude, exclude_d
...
       long_description_content_type='text/markdown',
       author='AssemblyLine',
       author_email='52798256+plocket@users.noreply.github.com',
       license='The MIT License (MIT)',
       url='https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/framework/altoolbox',
       packages=find_packages(),
       namespace_packages=['docassemble'],
-      install_requires=['holidays>=0.14.2', 'pandas>=1.4.2'],
+      install_requires=['pandas>=1.5.3'],
```
Comment (plocket) [inline setup.py:57]:
```suggestion
      install_requires=['holidays>=0.14.2', 'pandas>=1.4.2'],
```

**Comment by nonprofittechy** [comment]
@miabonardi ran into an interesting bug with the ThreePartsDate input on UpToCode. Not clear if it's related to the custom CSS on that site although I haven't experienced it myself so might be browser or screen size specific. See: https://github.com/nonprofittechy/docassemble-HousingCodeChecklist/issues/363

Perhaps we need to add an appropriate minimum `em` width to each field

**Comment by nonprofittechy** [comment]
Before I forget it, on call Plocket had suggested naming this "split_part_date". I'm not a big fan of changing the name but at least NNGroup uses that label for a date field with 3 drop down menus and discourages its use: https://www.nngroup.com/articles/date-input/

**Comment by miabonardi** [comment]
> @miabonardi ran into an interesting bug with the ThreePartsDate input on UpToCode. Not clear if it's related to the custom CSS on that site although I haven't experienced it myself so might be browser or screen size specific. See: [nonprofittechy/docassemble-HousingCodeChecklist#363](https://github.com/nonprofittechy/docassemble-HousingCodeChecklist/issues/363)
> 
> Perhaps we need to add an appropriate minimum `em` width to each field

Screen recording available - glad to share!

**Comment by plocket** [comment]
> Before I forget it, on call Plocket had suggested naming this "split_part_date". I'm not a big fan of changing the name but at least NNGroup uses that label for a date field with 3 drop down menus and discourages its use: https://www.nngroup.com/articles/date-input/

We were discussing renaming the date parts to use `AL` at the start of the names. If we decide to do it, maybe we could do it then.

> @miabonardi ran into an interesting bug with the ThreePartsDate input on UpToCode. Not clear if it's related to the custom CSS on that site

Good eye. It was CSS for that site specifically and I added a comment in the issue about it 👍

**Comment by plocket** [comment]
Ready for re-review? Not sure how multi-select got in here and not sure whether to delete it or not. Swear I made a new project and pulled from this branch. And didn't add the file, of course. It doesn't look like the version that's on `main`.

**Comment by plocket** [comment]
Also, this should be working with the latest version of docassemble, which fixed the bug! Huzzah!

**Review by nonprofittechy — COMMENTED** [review]
I didn't test to see if the last bug was fixed, but looking great--just need to make the names consistent throughout. We should keep the ThreePartsDate name.

File: docassemble/ALToolbox/data/static/al_split_date.css [hunk 2]
```diff
@@ -0,0 +1,28 @@
+/* ThreePartDate and BirthDate */
+
+/* More room for month name text. */
+.al_split_date .col-year {
```
[the remark this answers is not in the document]
Comment (plocket) [reply docassemble/ALToolbox/data/static/al_split_date.css:4]: I can change the class naming style. One addition perhaps - I recall us deciding that our classes, at least for containers, would start with `al` to keep them unique.

File: docassemble/ALToolbox/data/static/al_split_date.css
[the remark this answers is not in the document]
Comment (plocket) [reply docassemble/ALToolbox/data/static/al_split_date.css:1]: Last I remember, we had agreed to move towards 'al' at the beginning of included/imported file names too. Is it only the includable .yml filenames that we're changing? I think it's probably not worth hashing out here, so just let me know what to do, but next time we get discuss it, maybe we can find a good place to record our points and conclusions - what we've rejected and accepted and why.

**Comment by plocket** [comment]
Let me know about the `al` prefix and then I can implement these changes.

File: docassemble/ALToolbox/data/static/al_split_date.css
[the remark this answers is not in the document]
Comment (nonprofittechy) [reply docassemble/ALToolbox/data/static/al_split_date.css:1]: The names need to be consistent for people to be able to work out which files, modules, and class names are connected. 

While using `al_` as a prefix is a good although somewhat cluttered rule to follow on new code, it does not make sense to rename just half of an existing project to that convention.

**Comment by nonprofittechy** [comment]
> Let me know about the `al` prefix and then I can implement these changes.

I think the `al_` prefix is fine, although probably unnecessary. But the files and classes should have one "core" name and stick with it--`ThreePartsDate` `three_parts_date` `three-parts-date` are all fine variations if they follow the convention of the language they are in, but it's probably also fine to have them all follow one convention. You can add or leave off the `al_` prefix except to the actual custom datatype.

I think most important to remove references to `split_date`. There's not a compelling reason to rename this and having three very similar custom datatypes for backwards compatibility will be confusing.

**Comment by plocket** [comment]
Also, what should I do about multiselect.yml?

**Comment by plocket** [comment]
If you look at the first runs of tests, you can the unit tests consistently get the error `ERROR: Could not install packages due to an OSError: [Errno 2] No such file or directory: '/tmp/tmpbsy65sw8/output.json'`. Rerunning seems to sometimes solve this. Is it something on our end?

**Comment by plocket** [comment]
Ready for re-review

**Review by nonprofittechy — APPROVED** [review]

**Comment by plocket** [comment]
Still failing. Any ideas, @BryceStevenWilley?

**Comment by BryceStevenWilley** [comment]
Definitely not something on our end, I've seen it before when pip has trouble installing a particular package, but I can't find anything useful in the logs or from a search. I won't be able to take a closer look at in until later today, but I can take on tryingnto fix it.

**Comment by nonprofittechy** [comment]
> Also, what should I do about multiselect.yml?

We'll need to fix in a separate PR. In command line git you could unstage the file but not in DA.

**Review by nonprofittechy — COMMENTED** [review]
Just check the pluralization

File: docassemble/ALToolbox/data/static/al_three_part_date.css
[the remark this answers is not in the document]
Comment (plocket) [reply docassemble/ALToolbox/data/static/al_three_part_date.css:1]: Missed that, sorry

**Comment by nonprofittechy** [comment]
I tried looking into the unit test failure and it does seem specific to this branch. Any more thoughts there?

The immediate prior dependency that it seems to fail after is `guess_language-spirit` which hasn't changed in 6 years. I did notice the original repository for it is gone but it is still on PyPI

**Review by BryceStevenWilley — APPROVED** [review]
Got the python tests to pass. I [deleted the pip cache](https://github.com/SuffolkLITLab/docassemble-ALToolbox/actions/caches) that the test uses (which didn't fix it the first run, but I'm not clear that it made a new cache that first run), and then tried to [log a lot more things in pip](https://github.com/SuffolkLITLab/ALActions/commit/d0106b725a318909623567015d0dd1852313ad73) so we could figure out what was wrong, but it seemed to fix itself then. I'll leave the pip logs on verbose for a bit in case the problem comes back after we merge.

Also couldn't reproduce that biggest javascript bug we had earlier, but only on DA 1.4.46 (could still reproduce it on 1.4.44, which is good). Everything LGTM!

**Pull request merged**